docs/dev/concepts.rst

```diff
@@ -0,0 +1,175 @@
+execution. Here's an example of an action that logs a message to the console:
+
+.. code-block:: javascript
+
+   export default class ConsoleLogAction {
+     constructor(normandy, recipe) {
+       this.normandy = normandy; // The driver object
+       this.recipe = recipe; // Contains recipe arguments and other data
+     }
+
+     // Actions should return a Promise when executed
+     async execute() {
+       this.normandy.log(this.recipe.arguments.message, 'info');
+     }
+   }
+
+   // registerAction is in the sandbox's global object
+   registerAction('console-log', ConsoleLogAction);
+
+.. _recipes:
+
+Recipes
+-------
+A recipe is a model that represents a single instance of an action we want to
+perform. Recipes contain three important pieces of information:
+
+1. Client filtering data that determines which users the recipe should be run
+   for. This takes the form of a JEXL_ expression that is evaluated on the
+   client and has access to several pieces of data about the client, such as the
+   data collected by Telemetry_.
+2. A pointer to the :ref:`action <actions>` that this recipe should perform when
+   run.
+3. The configuration data to pass to the Action code when it is executed, called
+   the recipe's **arguments**.
+
+Recipes are stored in the Normandy database and the :ref:`runtime <runtime>`
+fetches them from the service during execution. Here's an example of a recipe in
+JSON format as returned by Normandy:
+
+.. code-block:: json
+
+   {
+      "id": 1,
+      "name": "Console Log Test",
+      "revision_id": 12,
+      "action": {
+         "name": "console-log",
+         "implementation_url": "https://self-repair.mozilla.org/api/v1/action/console-log/implementation/8ee8e7621fc08574f854972ee77be2a5280fb546/",
+         "arguments_schema": {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "description": "Log a message to the console",
+            "type": "object",
+            "properties": {
+               "message": {
+                  "default": "",
+                  "description": "Message to log to the console",
+                  "type": "string",
+               }
+            },
+            "required": [
+               "message"
+            ],
+         }
+      },
+      "arguments": {
+         "message": "It works!"
+      }
+   }
+
+Note that the recipe as returned by the API also contains data about the action,
+including a schema describing the arguments field.
+
+.. _JEXL: https://github.com/TechnologyAdvice/Jexl
+.. _Telemetry: https://wiki.mozilla.org/Telemetry
+
+.. _runtime:
+
+Runtime
+-------
+The runtime is an execution environment that can affect a running instance of
+Firefox. Upon activation (typically a few moments after Firefox launches), the
+runtime:
+
+1. Downloads :ref:`recipes <recipes>` from the Normandy service.
+2. Evaluates the recipe filters and filters out recipes that do not match the
+   client the runtime is installed within.
```

Comment:
To prevent fraudulent injections in the filtering logic, we should verify signatures prior to evaluating the filters.